nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/security/RetentionExpiredAction.java

```diff
@@ -0,0 +1,133 @@
+import static org.nuxeo.lib.stream.computation.AbstractComputation.OUTPUT_1;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.DocumentRef;
+import org.nuxeo.ecm.core.api.IdRef;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.PropertyException;
+import org.nuxeo.ecm.core.api.event.CoreEventConstants;
+import org.nuxeo.ecm.core.api.event.DocumentEventCategories;
+import org.nuxeo.ecm.core.api.event.DocumentEventTypes;
+import org.nuxeo.ecm.core.bulk.action.computation.AbstractBulkComputation;
+import org.nuxeo.ecm.core.event.Event;
+import org.nuxeo.ecm.core.event.EventService;
+import org.nuxeo.ecm.core.event.impl.DocumentEventContext;
+import org.nuxeo.lib.stream.computation.Topology;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.stream.StreamProcessorTopology;
+
+/**
+ * Bulk action executed for documents whose retention has expired.
+ *
+ * @since 11.1
+ */
+public class RetentionExpiredAction implements StreamProcessorTopology {
+
+    public static final String ACTION_NAME = "retentionExpired";
+
+    @Override
+    public Topology getTopology(Map<String, String> options) {
+        return Topology.builder()
+                       .addComputation(RetentionExpiredComputation::new, //
+                               Arrays.asList(INPUT_1 + ":" + ACTION_NAME, OUTPUT_1 + ":" + STATUS_STREAM))
+                       .build();
+    }
+
+    public static class RetentionExpiredComputation extends AbstractBulkComputation {
+
+        private static final Logger log = LogManager.getLogger(RetentionExpiredComputation.class);
+
+        public RetentionExpiredComputation() {
+            super(ACTION_NAME);
+        }
+
+        @Override
+        protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+            Collection<DocumentRef> refs = ids.stream().map(IdRef::new).collect(Collectors.toList());
+            for (DocumentRef ref : refs) {
+                // sanity checks
+                if (!session.isRecord(ref)) {
+                    log.debug("Document {} is not a record", ref);
+                    continue;
+                }
+                Calendar retainUntil = session.getRetentionUntil(ref);
+                if (retainUntil == null) {
+                    log.debug("Document {} was not under retention", ref);
+                    continue;
+                }
+                if (isUnderRetention(retainUntil)) {
+                    log.debug("Document {} is still under retention: {}", ref, retainUntil.toInstant());
```

Comment:
`toInstant` does things, could be interesting to change to lambda.

Reply:
Could you add `:` just after `Document`. We might need to add it ton conventions 🤔.